File: contrib/format101/python/src/format101/codec6.py
def compact_6bit_text_to_octets(text: bytes) -> bytes:
    """
    Reassemble 6-bit words (stored in the lower 6 bits of each byte) as 8-bit octets.

    This is based on a reference implementation found in the ecosystem around MAWSbin.
    """
    octets: list[int] = []
    buf = 0
    rss = [0, 4, 0, 2, 0, 0]  # right-shift amounts
    lss = [2, 0, 4, 0, 6, 0]  # left-shift amounts
    bcs = [6, 2, 4, 4, 2, 6]  # bits put in byte amounts
    i = 0  # step counter
    oc = 0  # input index
    bc = 0  # bits accumulated from 6-bit words
    while oc < len(text):
        bc += bcs[i]
        buf |= (((text[oc] & 63) >> rss[i]) << lss[i]) & 255
        if i in (1, 3, 5):
            octets.append(buf)
            buf = 0
        i += 1
        if i == 6:
            i = 0
        if bc == 6:
            bc = 0
            oc += 1
    return bytes(octets)


def expand_6bit_text_to_octets(text: bytes) -> bytes:
    """
    Decode TurboWin+ half-compressed payload bytes into the full 8-bit octet stream.

    TurboWin+'s encoder produces a payload where each byte is in the range 0x40..0x7F
    and encodes a 6-bit value in the lower bits, i.e.:

        six = (byte - 0x40) & 0x3F
    """
    return compact_6bit_text_to_octets(bytes(((b - 0x40) & 0x3F) for b in text))


def decode_turbowin_text_to_octets(text: bytes) -> bytes:
    """
    Decode TurboWin+ half-compressed payload bytes into the full 8-bit octet stream.

    Alias for expand_6bit_text_to_octets(), kept for clearer naming.
    """
    return expand_6bit_text_to_octets(text)


def encode_octets_to_6bit_words(octets: bytes) -> bytes:
    """
    Inverse of compact_6bit_text_to_octets(): reconstruct the original 6-bit words
    from octets, so that:

        compact_6bit_text_to_octets(encode_octets_to_6bit_words(octets)) == octets

    This implements the inverse mapping of the 6-step packing pattern used by the
    ecosystem reference decoder. It yields a deterministic 6-bit word stream.
    """
    if not octets:
        return b""

    out = bytearray()

    # For every 3 octets, we can reconstruct 4 6-bit words.
    # Derivation follows the compacting steps:
    #   o0 uses bits from w0 and w1
    #   o1 uses bits from w1 and w2
    #   o2 uses bits from w2 and w3
    i = 0
    n = len(octets)
    while i + 3 <= n:
        o0 = octets[i] & 0xFF
        o1 = octets[i + 1] & 0xFF
        o2 = octets[i + 2] & 0xFF

        w0 = (o0 >> 2) & 0x3F
        w1 = ((o0 & 0x03) << 4) | ((o1 >> 4) & 0x0F)
        w2 = ((o1 & 0x0F) << 2) | ((o2 >> 6) & 0x03)
        w3 = o2 & 0x3F

        out.extend((w0, w1, w2, w3))
        i += 3

    # Handle remaining octets (1 or 2). These correspond to 2 or 3 6-bit words.
    rem = n - i
    if rem == 1:
        o0 = octets[i] & 0xFF
        w0 = (o0 >> 2) & 0x3F
        w1 = (o0 & 0x03) << 4  # remaining 4 bits unknown -> 0
        out.extend((w0, w1))
    elif rem == 2:
        o0 = octets[i] & 0xFF
        o1 = octets[i + 1] & 0xFF
        w0 = (o0 >> 2) & 0x3F
        w1 = ((o0 & 0x03) << 4) | ((o1 >> 4) & 0x0F)
        w2 = (o1 & 0x0F) << 2  # remaining 2 bits unknown -> 0
        out.extend((w0, w1, w2))

    return bytes(out)


def encode_payload_octets_to_turbowin_text(octets: bytes) -> bytes:
    """
    Encode payload octets into TurboWin+ half-compressed text bytes (0x40..0x7F).

    This uses the inverse of the ecosystem compaction algorithm (compact_6bit_text_to_octets),
    yielding a deterministic 6-bit word stream, then maps each 6-bit word to a text byte
    by adding 0x40.
    """
    words = encode_octets_to_6bit_words(octets)
    return bytes(((w & 0x3F) + 0x40) for w in words)

File: contrib/format101/java/src/main/java/format101/Codec6.java
package format101;

import java.nio.charset.StandardCharsets;

public final class Codec6 {

    private Codec6() {}

    public static byte[] compact6BitWordsToOctets(byte[] sixBitWords) {
        java.util.ArrayList<Byte> octets = new java.util.ArrayList<>();
        int buf = 0;
        int[] rss = {0, 4, 0, 2, 0, 0};
        int[] lss = {2, 0, 4, 0, 6, 0};
        int[] bcs = {6, 2, 4, 4, 2, 6};
        int i = 0;
        int oc = 0;
        int bc = 0;
        while (oc < sixBitWords.length) {
            bc += bcs[i];
            int b = sixBitWords[oc] & 0x3F;
            buf |= (((b >> rss[i]) << lss[i]) & 0xFF);
            if (i == 1 || i == 3 || i == 5) {
                octets.add((byte) buf);
                buf = 0;
            }
            i++;
            if (i == 6) {
                i = 0;
            }
            if (bc == 6) {
                bc = 0;
                oc++;
            }
        }
        byte[] out = new byte[octets.size()];
        for (int j = 0; j < octets.size(); j++) {
            out[j] = octets.get(j);
        }
        return out;
    }

    public static byte[] decodeTurboWinTextToOctets(byte[] payloadText) {
        byte[] six = new byte[payloadText.length];
        for (int i = 0; i < payloadText.length; i++) {
            int b = payloadText[i] & 0xFF;
            six[i] = (byte) (((b - 0x40) & 0x3F));
        }
        return compact6BitWordsToOctets(six);
    }

    public static byte[] encodeOctetsTo6BitWords(byte[] octets) {
        if (octets.length == 0) {
            return new byte[0];
        }

        byte[] out = new byte[((octets.length / 3) * 4) + (octets.length % 3 == 0 ? 0 : (octets.length % 3 == 1 ? 2 : 3))];

        int oi = 0;
        int wi = 0;

        while (oi + 3 <= octets.length) {
            int o0 = octets[oi] & 0xFF;
            int o1 = octets[oi + 1] & 0xFF;
            int o2 = octets[oi + 2] & 0xFF;

            int w0 = (o0 >> 2) & 0x3F;
            int w1 = ((o0 & 0x03) << 4) | ((o1 >> 4) & 0x0F);
            int w2 = ((o1 & 0x0F) << 2) | ((o2 >> 6) & 0x03);
            int w3 = o2 & 0x3F;

            out[wi++] = (byte) w0;
            out[wi++] = (byte) w1;
            out[wi++] = (byte) w2;
            out[wi++] = (byte) w3;

            oi += 3;
        }

        int rem = octets.length - oi;
        if (rem == 1) {
            int o0 = octets[oi] & 0xFF;
            int w0 = (o0 >> 2) & 0x3F;
            int w1 = ((o0 & 0x03) << 4);
            out[wi++] = (byte) w0;
            out[wi++] = (byte) w1;
        } else if (rem == 2) {
            int o0 = octets[oi] & 0xFF;
            int o1 = octets[oi + 1] & 0xFF;
            int w0 = (o0 >> 2) & 0x3F;
            int w1 = ((o0 & 0x03) << 4) | ((o1 >> 4) & 0x0F);
            int w2 = ((o1 & 0x0F) << 2);
            out[wi++] = (byte) w0;
            out[wi++] = (byte) w1;
            out[wi++] = (byte) w2;
        }

        return out;
    }

    public static byte[] encodePayloadOctetsToTurboWinText(byte[] octets) {
        byte[] words = encodeOctetsTo6BitWords(octets);
        byte[] out = new byte[words.length];
        for (int i = 0; i < words.length; i++) {
            out[i] = (byte) ((words[i] & 0x3F) + 0x40);
        }
        return out;
    }

    public static String latin1String(byte[] bytes) {
        return new String(bytes, StandardCharsets.ISO_8859_1);
    }
}
